### PR TITLE
Otp feature

### PR DESCRIPTION
**Pull request by XshubhamX**
OTP functionality, under development #245

### CODE REVIEW

**Comment by XshubhamX** [comment]
@palisadoes sir OTP feature is ready for review.

File: helper_functions/confirmOtp.js [hunk 1]
```diff
@@ -0,0 +1,24 @@
+const Otp = require('../models/Otp');
+
+const confirmOtp = async (email, otp) => {
+  let otp_object;
```
Comment (impiyush83) [inline helper_functions/confirmOtp.js:4]: Use const

File: helper_functions/confirmOtp.js [hunk 2]
```diff
@@ -0,0 +1,24 @@
...
+
+  try {
+    otp_object = await Otp.findOne({
+      email,
+      text: otp,
+    });
+  } catch (e) {
+    throw new Error(e);
+  }
+
```
Comment (impiyush83) [inline helper_functions/confirmOtp.js:16]: Dont use try catch here, use in the outer function

File: resolvers/functions/sendMail.js [hunk 3]
```diff
@@ -0,0 +1,34 @@
+const sgMail = require('@sendgrid/mail');
+const otp = require('otp-generator');
+
+const gridKey = process.env.GRID_KEY;
+
+sgMail.setApiKey(gridKey);
+
+let required_otp;
```
Comment (impiyush83) [inline resolvers/functions/sendMail.js:8]: dont declare variables   globally

File: resolvers/functions/sendMail.js [hunk 4]
```diff
@@ -0,0 +1,34 @@
...
+
+const sendConfirmationEmail = async (email) => {
+  //generating otp
+
+  try {
+    required_otp = otp.generate(6, {
```
Comment (impiyush83) [inline resolvers/functions/sendMail.js:14]: use const here

**Review by impiyush83 — CHANGES_REQUESTED** [review]

File: resolvers/functions/sendMail.js (hunk 3 above)
Comment (XshubhamX) [reply resolvers/functions/sendMail.js:8]: 🤘

File: resolvers/functions/sendMail.js (hunk 4 above)
Comment (XshubhamX) [reply resolvers/functions/sendMail.js:14]: 🤘

**Comment by XshubhamX** [comment]
@palisadoes sir this pr is ready for review.

File: helper_functions/confirmOtp.js (hunk 2 above)
Comment (XshubhamX) [reply helper_functions/confirmOtp.js:16]: @impiyush83 can you elaborate?

Comment (impiyush83) [reply helper_functions/confirmOtp.js:16]: @XshubhamX  Its no need to put try-catch in the child/helper function. You just throw an error from helper functions and let the handler function/function who is calling this confirmOtp have the try-catch block.

Comment (XshubhamX) [reply helper_functions/confirmOtp.js:16]: ok got it

File: resolvers/auth_mutations/signup.js [hunk 5]
```diff
@@ -8,14 +8,12 @@ const {
...
 
 module.exports = async (parent, args) => {
-  const emailTaken = await User.findOne({
-    email: args.data.email.toLowerCase(),
-  });
-  if (emailTaken) {
-    throw new Error('Email address taken.');
-  }
+  //Checking if Otp is valid
+
```
[the remark this answers is not in the document]
Comment (XshubhamX) [reply resolvers/auth_mutations/signup.js:15]: ok

**Comment by palisadoes** [comment]
@XshubhamX Is this specifically for email verification and/or text messages?

My concerns with using OTP are:

1. It adds to the steps required to get the application setup
2. It assumes that the organization has funds to pay for the service. 
3. It assumes persons have an email address. Many don't, or don't remember what they used to get things setup. I have experienced this in the past.

These are issues that can be overcome, especially if the user has the option to implement OTP at their choosing.

I have these questions:

1. What OTP various options are you proposing?
2. How will the `talawa-admin` user select between the various OTP methods?
2. How will this be configured using `talawa-admin`? There needs to be a way to do this simply. Manually adding it to the API configuration is not user friendly. @yasharth291 is working on a number of related features. You should talk with him. This should not be done in isolation.
3. Is there a modular / plugin way in which this can be done where the user can select the OTP provider? Do you think this approach is advisable?
2. Is there an option to not select OTP at all?
2. How will this feature integrate with this [priority feature](https://palisadoesfoundation.github.io/talawa-docs/docs/features/members/members-preapprovals)?

**Review by palisadoes — CHANGES_REQUESTED** [review]
@XshubhamX Is this specifically for email verification and/or text messages?

My concerns with using OTP are:

1. It adds to the steps required to get the application setup
2. It assumes that the organization has funds to pay for the service. 
3. It assumes persons have an email address. Many don't, or don't remember what they used to get things setup. I have experienced this in the past.

These are issues that can be overcome, especially if the user has the option to implement OTP at their choosing.

I have these questions:

1. What OTP various options are you proposing?
2. How will the `talawa-admin` user select between the various OTP methods?
2. How will this be configured using `talawa-admin`? There needs to be a way to do this simply. Manually adding it to the API configuration is not user friendly. @yasharth291 is working on a number of related features. You should talk with him. This should not be done in isolation.
3. Is there a modular / plugin way in which this can be done where the user can select the OTP provider? Do you think this approach is advisable?
2. Is there an option to not select OTP at all?
2. How will this feature integrate with this [priority feature](https://palisadoesfoundation.github.io/talawa-docs/docs/features/members/members-preapprovals)?

**Review by jasbirrajrana — COMMENTED** [review]
**LGTM**

**Review by ranilwallace — APPROVED** [review]
LGTM 👍

**Comment by ranilwallace** [comment]
@XshubhamX Please resolve all conversations so we can merge

**Comment by palisadoes** [comment]
@XshubhamX We'd like to proceed with this, but we need the questions answered.

**Comment by palisadoes** [comment]
@XshubhamX  As I haven't heard back from you. I'm going to close PR #248 and #245. We prefer to go with an email service for the reasons stated previously.

**Pull request closed without merging**